### PR TITLE
Compile-time evaluate constant BitArray int segments on JavaScript

### PR DESCRIPTION
**Pull request by richard-viney**
Implements https://github.com/gleam-lang/gleam/discussions/3724.

## Details

1. `Int`s in bit array expressions with a known value and size at compile time are now evaluated to bytes instead of using `sizedInt()` at runtime.

2. `Int`s in bit array patterns with a known value and size at compile time are now evaluated then matched byte-wise instead of using `intFromSlice()` at runtime.

3. The above is currently limited to `Int` sizes <= 48 bits in order to be under the JS safe integer limit, because otherwise:
   1. Code such as `<<0:8192>>` would evaluate to `[0,0,0,0, ...]`to 1024 entries, which seems undesirable (though I'm not sure if such code is commonly encountered).
   2. Code such as `let assert <<0xFFFFFFFFFFFFFFFF:64>> == <<255, 255, 255, 255, 255, 255, 255, 255>>` would change behaviour, albeit in a correct direction that would match Erlang. This may be desirable, but I've taken the conservative route initially.
   3. See discussion https://github.com/gleam-lang/gleam/discussions/3724#discussioncomment-11013077 for more details.

## Performance

Evaluating `<<-1:32>>`: 10-11x faster.

Evaluating a large lookup table of 16-bit integers: 26x faster.

Pattern matching a 32-bit signed integer: 2.7-7.1x faster, depending on how early a non-matching byte occurs.

Generated JavaScript code size is up to ~70% smaller. There are a few cases when using 40-bit and 48-bit wide integers that can see code size increases of ~20% with certain values, but the average case still sees a code size reduction on the wider integers.

### CODE REVIEW

File: compiler-core/src/javascript/expression.rs [hunk 1]
```diff
@@ -1485,8 +1518,8 @@ fn bit_array<'a>(
 
 #[derive(Debug)]
 struct SizedBitArraySegmentDetails<'a> {
-    has_explicit_size: bool,
     size: Document<'a>,
+    size_value: Option<BigInt>, // This is set when the segment's size is known at compile time
```
Comment (lpil) [inline compiler-core/src/javascript/expression.rs:1522]: Documentation should be on the line above and use `///` please

File: compiler-core/src/javascript/tests/bit_arrays.rs [hunk 2]
```diff
@@ -122,11 +122,27 @@ fn go() {
 }
 
 #[test]
-fn sized_big_endian() {
+fn sized_with_compile_time_evaluation() {
```
Comment (lpil) [inline compiler-core/src/javascript/tests/bit_arrays.rs:125]: Undo the changes to existing files. Tests should be added to, not deleted or edited.

File: compiler-core/src/javascript/tests/bit_arrays.rs [hunk 3]
```diff
@@ -122,11 +122,27 @@ fn go() {
...
     assert_js!(
         r#"
 fn go() {
+  <<0:8>>
+  <<4000:16>>
+  <<80_000:16>>
+  <<-80_000:16>>
+  <<-1:48>>
```
Comment (lpil) [inline compiler-core/src/javascript/tests/bit_arrays.rs:133]: Only test one thing per test please.

File: compiler-core/Cargo.toml [hunk 4]
```diff
@@ -46,6 +46,7 @@ unicode-segmentation = "1.12.0"
 bimap = "0.6.3"
 # Parsing of arbitrary width int values
 num-bigint = "0.4.6"
+num-traits = "0.2.19"
```
[the remark this answers is not in the document]
Comment (richard-viney) [reply compiler-core/Cargo.toml:49]: The most important thing it's for is the `ToPrimitive` trait that provides `BigInt::to_u8()` and `BigInt::to_u32()`, i.e. conversions to primitive integer types. The `num-traits` crate is widely used and is an existing transitive dependency, hence it seemed reasonable to add.

Alternatively, it's possible to implement the conversion ourselves from the data held inside a `BigInt`, which would avoid adding the new explicit dependency:

```rust
fn bigint_to_u32(i: &BigInt) -> Option<u32> {
    match i.sign() {
        num_bigint::Sign::Minus => None,
        num_bigint::Sign::NoSign => Some(0),
        num_bigint::Sign::Plus => {
            if i > &BigInt::from(u32::MAX) {
                None
            } else {
                i.iter_u32_digits().next()
            }
        }
    }
}
```

File: compiler-core/src/javascript.rs [hunk 5]
```diff
@@ -810,3 +813,56 @@ fn bool(bool: bool) -> Document<'static> {
...
+    location: SrcSpan,
+) -> Result<Vec<u8>, Error> {
+    // Clamp negative sizes to zero
+    let size = size.max(BigInt::ZERO);
+
+    // The segment size in bits is not allowed to exceed the range of a u32. At runtime the
+    // limit is lower than this and depends on the JS engine. V8's limit is currently
+    // 2^30-1 bits.
+    let size = size.to_u32().ok_or_else(|| Error::Unsupported {
+        feature: "Integer segment size greater than 2^32-1".into(),
```
[the remark this answers is not in the document]
Comment (richard-viney) [reply compiler-core/src/javascript.rs:842]: It would be breaking if this function was called with integers of any size, but it's limited to those up to `SAFE_INT_SEGMENT_MAX_SIZE` as ints of arbitrary size would result in enormous generated code from input such as `<<0:1_000_000>>`.

The comment and error handling here was misleading and has been fixed.

File: compiler-core/src/javascript.rs [hunk 6]
```diff
@@ -810,3 +813,56 @@ fn bool(bool: bool) -> Document<'static> {
...
+
+    let byte_mask = BigInt::from(0xFF);
+
+    // Convert value to the desired number of bytes
+    let mut bytes = vec![0u8; size as usize / 8];
+    for byte in bytes.iter_mut() {
+        *byte = (&value & &byte_mask)
+            .to_u8()
+            .expect("bitwise and result to be a u8");
+        value >>= 8;
```
[the remark this answers is not in the document]
Comment (richard-viney) [reply compiler-core/src/javascript.rs:860]: Changed to `BigInt` division.

File: compiler-core/src/javascript.rs [hunk 7]
```diff
@@ -810,3 +813,56 @@ fn bool(bool: bool) -> Document<'static> {
...
+        let value_modulus = BigInt::one() << size;
+        value = &value_modulus + (value % &value_modulus);
+    }
+
+    let byte_mask = BigInt::from(0xFF);
+
+    // Convert value to the desired number of bytes
+    let mut bytes = vec![0u8; size as usize / 8];
+    for byte in bytes.iter_mut() {
+        *byte = (&value & &byte_mask)
```
[the remark this answers is not in the document]
Comment (richard-viney) [reply compiler-core/src/javascript.rs:857]: Changed to `BigInt` modulo.

File: compiler-core/src/javascript.rs [hunk 8]
```diff
@@ -810,3 +813,56 @@ fn bool(bool: bool) -> Document<'static> {
...
+    // limit is lower than this and depends on the JS engine. V8's limit is currently
+    // 2^30-1 bits.
+    let size = size.to_u32().ok_or_else(|| Error::Unsupported {
+        feature: "Integer segment size greater than 2^32-1".into(),
+        location,
+    })?;
+
+    // Convert negative number to two's complement representation
+    if value < BigInt::ZERO {
+        let value_modulus = BigInt::one() << size;
```
[the remark this answers is not in the document]
Comment (richard-viney) [reply compiler-core/src/javascript.rs:848]: Changed to `BigInt::pow()`.

File: compiler-core/src/javascript/expression.rs (hunk 1 above)
Comment (richard-viney) [reply compiler-core/src/javascript/expression.rs:1522]: Fixed.

File: compiler-core/src/javascript.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/src/javascript.rs]: Thank you

File: compiler-core/Cargo.toml (hunk 4 above)
[the remark this answers is not in the document]
Comment (lpil) [reply compiler-core/Cargo.toml:49]: OK, let's stick with this. Thank you

File: compiler-core/src/javascript/tests/bit_arrays.rs (hunk 3 above)
Comment (richard-viney) [reply compiler-core/src/javascript/tests/bit_arrays.rs:162]: Fixed.

**Review by lpil — COMMENTED** [review]
Fantastic! Thank you!

Seeing as this is tricky runtime behaviour could you add some integrations tests please 🙏  

https://github.com/gleam-lang/gleam/blob/31966a0b3503d9429cc016d29b1ba88569e22fc6/test/language/test/language_test.gleam#L954

**Comment by richard-viney** [comment]
I've updated the existing language tests so each test is run both as a constant value that is compile-time evaluated, and as a dynamic value that is runtime evaluated.

**Review by lpil — APPROVED** [review]
Amazing work again!!! 💜
